filogic: add support for ASUS TUF AX6000

patrykk opened the pull request:
### **Hardware specification:**

- SOC: MediaTek MT7986
- RAM: 512MB DDR3
- FLASH: 256MB SPI-NAND
- WIFI: Mediatek MT7986 DBDC 802.11ax 2.4/5 GHz 4T4R
- ETH: MediaTek MT7530 Switch (LAN)
- MaxLinear GPY211C 2.5 N-Base-T PHY (WAN)
- MaxLinear GPY211C 2.5 N-Base-T PHY (LAN)
- USB: 1 x USB 3.1
- UART: 3V3 115200 8N1 (Do not connect VCC)

### **Installation**

1. Download the OpenWrt initramfs image. Copy the image to a TFTP server
reachable at 192.168.1.70/24. Rename the image to TUF-AX6000.bin.

2. Connect to the serial console, interrupt the auto boot process by pressing '4' when prompted or press '1' and set
 client IP, server IP and name of the image.

( We don't need to open the case or even soldering anything. We can use three goldpin wires, remove from them plastic cover from connectors and connect to the console via the case holes. We can see three console holes ( from down RX, TX, Ground - partially covered).

3. Download & Boot the OpenWrt initramfs image.

When pressed '4'

     $ setenv ipaddr 192.168.1.1
     $ setenv serverip 192.168.1.70
     $ tftpboot 0x46000000 TUF-AX6000.bin
     $ bootm 0x46000000

When pressed '1' just write image name like below:
    
     1: Load System code to SDRAM via TFTP.
     Please Input new ones /or Ctrl-C to discard
     Input device IP (192.168.1.1) ==:
     Input server IP (192.168.1.70) ==:
     Input Linux Kernel filename (TUF-AX6000.trx) ==:

4. Wait for OpenWrt to boot. Transfer the sysupgrade image to the device using scp 
and install using sysupgrade.

    $ sysupgrade -n <path-to-sysupgrade.bin>

### **Missing features**

- The 2.5Gb LAN port LED is ON during boot or when the LAN cable is disconnected
- The cover yellow light is not supported. (only blue one)

blocktrron commented on target/linux/mediatek/filogic/base-files/lib/preinit/10_fix_eth_mac.sh:
```diff
@@ -16,6 +16,12 @@ preinit_set_mac_address() {
 		ip link set dev eth0 address "$addr"
 		ip link set dev eth1 address "$addr"
 		;;
+	asus,tuf-ax6000)
+		CI_UBIPART="UBI_DEV"
+		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
+		ip link set dev eth0 address "$addr"
+		ip link set dev eth1 address "$addr"
+		;;
```
Combine with ax4200

blocktrron commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -61,6 +61,11 @@ platform_do_upgrade() {
 		CI_KERNPART="linux"
 		nand_do_upgrade "$1"
 		;;
+	asus,tuf-ax6000)
+		CI_UBIPART="UBI_DEV"
+		CI_KERNPART="linux"
+		nand_do_upgrade "$1"
+		;;
```
Combine with ax4200

blocktrron commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -25,6 +25,15 @@ case "$board" in
 		[ "$PHYNBR" = "0" ] && macaddr_setbit_la $(macaddr_add $addr 1) > /sys${DEVPATH}/macaddress
 		[ "$PHYNBR" = "1" ] && echo "$addr" > /sys${DEVPATH}/macaddress
 		;;
+	asus,tuf-ax6000)
+		CI_UBIPART="UBI_DEV"
+		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
+		# Originally, phy0 is phy1 mac with LA bit set. However, this would conflict
+		# addresses on multiple VIFs with the other radio. Set LA bit and increment
+		# mac-address instead.
+		[ "$PHYNBR" = "0" ] && macaddr_setbit_la $(macaddr_add $addr 1) > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && echo "$addr" > /sys${DEVPATH}/macaddress
+		;;
 	bananapi,bpi-r3)
```
Combine with ax4200

blocktrron commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/firmware/11-mt76-caldata:
```diff
@@ -27,6 +27,10 @@ case "$FIRMWARE" in
 		CI_UBIPART="UBI_DEV"
 		caldata_extract_ubi "Factory" 0x0 0x1000
 		;;
+	asus,tuf-ax6000)
+		CI_UBIPART="UBI_DEV"
+		caldata_extract_ubi "Factory" 0x0 0x1000
+		;;
```
Combine with ax4200

blocktrron commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -80,6 +83,12 @@ mediatek_setup_macs()
 		wan_mac="${addr}"
 		lan_mac="${addr}"
 		;;
+	asus,tuf-ax6000)
+		CI_UBIPART="UBI_DEV"
+		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
+		wan_mac="${addr}"
+		lan_mac="${addr}"
+		;;
```
Combine with ax4200

blocktrron commented on target/linux/mediatek/dts/mt7986a-asus-tuf-ax6000.dts:
```diff
@@ -0,0 +1,382 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+
+/dts-v1/;
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
```
Judging from pictures on Google photos, this is a different base board than the ax4200. I've had a look, since my ax4200 came with an unpopulated area for a second 2.5GE PHY.

patrykk replied:
base board looks different, that's true. What do you suggest?

blocktrron replied:
No suggestion, just an observation. I've thought this comment was separate from the review.

patrykk replied:
> No suggestion, just an observation. I've thought this comment was separate from the review.

Ok. This router have a different case and main board than ax4200. I have both of them and I recommend to buy the ax6000 version.

patrykk committed:
filogic: add support for ASUS TUF AX6000

Hardware

 SOC: MediaTek MT7986
 RAM: 512MB DDR3
 FLASH: 256MB SPI-NAND
 WIFI: Mediatek MT7986 DBDC 802.11ax 2.4/5 GHz 4T4R
 ETH: MediaTek MT7530 Switch (LAN)
 MaxLinear GPY211C 2.5 N-Base-T PHY (WAN)
 MaxLinear GPY211C 2.5 N-Base-T PHY (LAN)
 UART: 3V3 115200 8N1 (Do not connect VCC)
 USB 1x3.1

Installation

     Download the OpenWrt initramfs image. Copy the image to a TFTP server
     reachable at 192.168.1.70/24. Rename the image to TUF-AX6000.bin.

     Connect to the serial console, interrupt the auto boot process by pressing '4' when prompted or press '1' and set
     client IP, server IP and name of the image.

 ( We don't need to open the case or even soldering anything. We can use three goldpin wires, remove from them plastic cover from connectors and connect to the console via the case holes. We can see three console holes ( from down RX, TX, Ground - partially covered).

     Download & Boot the OpenWrt initramfs image.

 When pressed '4'
 $ setenv ipaddr 192.168.1.1
 $ setenv serverip 192.168.1.70
 $ tftpboot 0x46000000 TUF-AX6000.bin
 $ bootm 0x46000000

 When pressed '1' just write image name like below:
 1: Load System code to SDRAM via TFTP.
 Please Input new ones /or Ctrl-C to discard
 Input device IP (192.168.1.1) ==:
 Input server IP (192.168.1.70) ==:
 Input Linux Kernel filename (TUF-AX6000.trx) ==:

     Wait for OpenWrt to boot. Transfer the sysupgrade image to the device
     using scp and install using sysupgrade.

     $ sysupgrade -n <path-to-sysupgrade.bin>

Missing features

     The 2.5Gb LAN port LED is ON during boot or when the LAN cable is disconnected
     The cover yellow light is not supported. (only blue one)

Signed-off-by: Patryk Kowalczyk <patryk@kowalczyk.ws>

blocktrron commented:
Merged to my staging tree - thanks!

blocktrron commented on target/linux/mediatek/dts/mt7986a-asus-tuf-ax6000.dts:
```diff
@@ -0,0 +1,382 @@
+		compatible = "ethernet-phy-ieee802.3-c45";
+		reg = <5>;
+
+		mxl,led-config = <0x03f0 0x0 0x0 0x0>;
```
Can you check if altering the LED Port configuration fixes the LED behavior on the LAN port? Documentation for these registers can be found in the GPY211 datasheet which is public accessible.

patrykk replied:
I'll try to investigate the topic.

patrykk replied:
Base on the documentation https://assets.maxlinear.com/web/documents/617810_gpy211b1vc_gpy211c0vc_ds_rev1.4.pdf , page 142, I think it should be 03f0. We can also try to add reset option for LED Pin 0  ( page 141) .

blocktrron replied:
I suspect the LEDs are connected to a different pin. The 4 32 bit integer values reflect a single LED per-value.

patrykk replied:
without this line the LED (only one for this separate GPY) lights up permanently when link is not connected and goes out when link is up.

blocktrron replied:
@patrykk You'll need to identify which LED pin on the PHY the LED is actually connected to. For this, i recommend to enable RX/TX activity indication (blinking) on each pin separately. When you've identified the used LED pin, you can determine the configuration of said pin.

patrykk replied:
@blocktrron I confirm the led is connected to pin 0 and best solution until now is 03f0.  In the no link status the led shining, when link is up the  led also shining and flashing in reverse during tx, rx activity.

basmevissen commented:
When are we going to see this in OpenWRT main please? Is there something holding back merging?

patrykk commented:
> When are we going to see this in OpenWRT main please? Is there something holding back merging?

@basmevissen  Please contact with @blocktrron. The mergning process is not fully clear for me

patrykk commented:
Hi, after sync with main I found problem with firmware for wifi. I need to introduce minor change.

patrykk commented:
@blocktrron  @basmevissen  I created new commit which include changes for wifi: https://github.com/openwrt/openwrt/compare/main...patrykk:openwrt:asus_tuf_ax6000 [ file changed: 11-mt76-caldata].

remittor commented:
@patrykk , @dangowrt , @blocktrron 
If, as a result of building the initramfs-image, its size exceeds 16 MiB, then when you try to load such a FIT-image, the bootloader will generate a critical error:
```
TUF-AX4200 bootloader version: 1.0.0.2
MAC Address: C8:7F:54:78:BB:3E
HwID A
Read 40 bytes from volume linux to 0000000046000000
   FIT/FDT format image found at 0x46000000,size 0x176e6d0
Read 24569592 bytes from volume linux to 0000000046000000
## Loading kernel from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'kernel-1' kernel subimage
     Description:  ARM64 OpenWrt Linux-5.15.134
     Type:         Kernel Image
     Compression:  lzma compressed
     Data Start:   0x460000ec
     Data Size:    3828481 Bytes = 3.7 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: 0x48000000
     Entry Point:  0x48000000
     Hash algo:    crc32
     Hash value:   97f8d931
     Hash algo:    sha1
     Hash value:   ef8eb4f46fb7374e60d946f54b8b7f6796967a1e
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Loading ramdisk from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'initrd-1' ramdisk subimage
     Description:  ARM64 OpenWrt asus_tuf-ax4200 initrd
     Type:         RAMDisk Image
     Compression:  Unknown Compression
     Data Start:   0x463a6d2c
     Data Size:    20718272 Bytes = 19.8 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: unavailable
     Entry Point:  unavailable
     Hash algo:    crc32
     Hash value:   b415a9b1
     Hash algo:    sha1
     Hash value:   2a81a603216028fc1698d550b70c4f56458793cf
   Verifying Hash Integrity ... crc32+ sha1+ OK
WARNING: 'compression' nodes for ramdisks are deprecated, please fix your .its file!
## Loading fdt from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'fdt-1' fdt subimage
     Description:  ARM64 OpenWrt asus_tuf-ax4200 device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x477690f8
     Data Size:    20668 Bytes = 20.2 KiB
     Architecture: AArch64
     Load Address: 0x47000000
     Hash algo:    crc32
     Hash value:   ab5f8197
     Hash algo:    sha1
     Hash value:   23933b076fe071709ca592f8d9e30dff7425a5f6
   Verifying Hash Integrity ... crc32+ sha1+ OK
Error: fdt overwritten
Could not find a valid device tree
```
After this, the bootloader will copy the kernel from another ubifs volume into memory.

The reason for this error is simple:
`0x46000000 + 16MiB + 1 = 0x47000001`

Solution:
Need in the file `filogic` delete the line: `DEVICE_DTS_LOADADDR := 0x47000000`

After this, the bootloader will not try to move `fdt-1` to another memory location.

All stock images from Asus do not contain `load` parameter for `fdt-1` !!!

dangowrt commented:
> All stock images from Asus do not contain `load` parameter for `fdt-1` !!!

Yes, relocating fdt is only needed for boards on which the bootloader has to apply device tree overlays, such as the BPi-R3.

melroy89 commented:
> @patrykk , @dangowrt , @blocktrron If, as a result of building the initramfs-image, its size exceeds 16 MiB, then when you try to load such a FIT-image, the bootloader will generate a critical error:
> 
> ```
> TUF-AX4200 bootloader version: 1.0.0.2
> MAC Address: C8:7F:54:78:BB:3E
> HwID A
> Read 40 bytes from volume linux to 0000000046000000
>    FIT/FDT format image found at 0x46000000,size 0x176e6d0
> Read 24569592 bytes from volume linux to 0000000046000000
> ## Loading kernel from FIT Image at 46000000 ...
>    Using 'config-1' configuration
>    Trying 'kernel-1' kernel subimage
>      Description:  ARM64 OpenWrt Linux-5.15.134
>      Type:         Kernel Image
>      Compression:  lzma compressed
>      Data Start:   0x460000ec
>      Data Size:    3828481 Bytes = 3.7 MiB
>      Architecture: AArch64
>      OS:           Linux
>      Load Address: 0x48000000
>      Entry Point:  0x48000000
>      Hash algo:    crc32
>      Hash value:   97f8d931
>      Hash algo:    sha1
>      Hash value:   ef8eb4f46fb7374e60d946f54b8b7f6796967a1e
>    Verifying Hash Integrity ... crc32+ sha1+ OK
> ## Loading ramdisk from FIT Image at 46000000 ...
>    Using 'config-1' configuration
>    Trying 'initrd-1' ramdisk subimage
>      Description:  ARM64 OpenWrt asus_tuf-ax4200 initrd
>      Type:         RAMDisk Image
>      Compression:  Unknown Compression
>      Data Start:   0x463a6d2c
>      Data Size:    20718272 Bytes = 19.8 MiB
>      Architecture: AArch64
>      OS:           Linux
>      Load Address: unavailable
>      Entry Point:  unavailable
>      Hash algo:    crc32
>      Hash value:   b415a9b1
>      Hash algo:    sha1
>      Hash value:   2a81a603216028fc1698d550b70c4f56458793cf
>    Verifying Hash Integrity ... crc32+ sha1+ OK
> WARNING: 'compression' nodes for ramdisks are deprecated, please fix your .its file!
> ## Loading fdt from FIT Image at 46000000 ...
>    Using 'config-1' configuration
>    Trying 'fdt-1' fdt subimage
>      Description:  ARM64 OpenWrt asus_tuf-ax4200 device tree blob
>      Type:         Flat Device Tree
>      Compression:  uncompressed
>      Data Start:   0x477690f8
>      Data Size:    20668 Bytes = 20.2 KiB
>      Architecture: AArch64
>      Load Address: 0x47000000
>      Hash algo:    crc32
>      Hash value:   ab5f8197
>      Hash algo:    sha1
>      Hash value:   23933b076fe071709ca592f8d9e30dff7425a5f6
>    Verifying Hash Integrity ... crc32+ sha1+ OK
> Error: fdt overwritten
> Could not find a valid device tree
> ```
> 
> After this, the bootloader will copy the kernel from another ubifs volume into memory.
> 
> The reason for this error is simple: `0x46000000 + 16MiB + 1 = 0x47000001`
> 
> Solution: Need in the file `filogic` delete the line: `DEVICE_DTS_LOADADDR := 0x47000000`
> 
> After this, the bootloader will not try to move `fdt-1` to another memory location.
> 
> All stock images from Asus do not contain `load` parameter for `fdt-1` !!!

This is a critical issue! Please consider merging those PRs.

dangowrt commented:
> All stock images from Asus do not contain `load` parameter for `fdt-1` !!!

Just filling the MAC address can be done in-place and doesn't need relocation. However, operations which increase the size of the DTB (such as applying overlays which add nodes or replace values with longer values) do.

Hence, loading the FDT is only needed on devices on which the bootloader needs to modify the device tree using device tree overlays. This is currently only the case for BananaPi router boards as well as MediaTek reference boards which allow multiple hardware configuration options to be used (e.g. different devices to boot from, or different setup of muxes for ethernet, PCIe or storage).

basmevissen commented:
And what about making sure there is enough room by modifying the DTS or FIT image load address? It is now 24MiB, so with 32MiB, it would have sufficient headroom.

remittor commented:
> so with 32MiB, it would have sufficient headroom.

The images I builded sometimes exceed 32 MB.

remittor commented:
@melroy89 , @dangowrt 
> This is a critical issue! Please consider merging those PRs.

Still not fixed.

@blocktrron , a new problem with the bootloader is described here:
https://forum.openwrt.org/t/205261
A solution is also described there.
